disperser/dataapi/nonsigner_utils.go

```diff
@@ -0,0 +1,182 @@
+func (oqi OperatorQuorumIntervals) GetQuorums(operatorId string, blockNum uint32) []uint8 {
+	quorums := make([]uint8, 0)
+	for q, intervals := range oqi[operatorId] {
+		// Note: if len(intervals) is large, we can perform binary search here.
+		// In practice it should be quite small given that the quorum change is
+		// not frequent, so search it with brute force here.
+		live := false
+		for _, interval := range intervals {
+			if interval.StartBlock > blockNum {
+				break
+			}
+			if blockNum <= interval.EndBlock {
+				live = true
+				break
+			}
+		}
+		if live {
+			quorums = append(quorums, q)
+		}
+	}
+	return quorums
+}
+
+// CreateOperatorQuorumIntervals creates OperatorQuorumIntervals that are within the
+// the block interval [startBlock, endBlock] for operators.
+//
+// The parameters:
+//   - startBlock, endBlock: specifying the block interval of interest.
+//     Requires: startBlock <= endBlock.
+//   - operatorInitialQuorum: the initial quorums at startBlock that operators were
+//     registered in.
+//     Requires: operatorInitialQuorum[op] is non-empty for each operator "op".
+//   - addedToQuorum, removedFromQuorum: a sequence of events that added/removed operators
+//     to/from quorums.
+//     Requires:
+//     1) the block numbers for all events are in range [startBlock+1, endBlock];
+//     2) the events are in ascending order by block number for each operator "op".
+func CreateOperatorQuorumIntervals(
+	startBlock uint32,
+	endBlock uint32,
+	operatorInitialQuorum map[string][]uint8,
+	addedToQuorum map[string][]*OperatorQuorum,
+	removedFromQuorum map[string][]*OperatorQuorum,
+) (OperatorQuorumIntervals, error) {
+	if startBlock > endBlock {
+		msg := "the endBlock must be no less than startBlock, but found " +
+			"startBlock: %d, endBlock: %d"
+		return nil, fmt.Errorf(msg, startBlock, endBlock)
+	}
+	operatorQuorumIntervals := make(OperatorQuorumIntervals)
+	addedToQuorumErr := "cannot add operator %s to quorum %d at block number %d, " +
+		"the operator is already in the quorum since block number %d"
+	for op, initialQuorums := range operatorInitialQuorum {
+		if len(initialQuorums) == 0 {
+			return nil, fmt.Errorf("operator %s must be in at least one quorum at block %d", op, startBlock)
+		}
+		operatorQuorumIntervals[op] = make(map[uint8][]BlockInterval)
+		openQuorum := make(map[uint8]uint32)
+		for _, q := range initialQuorums {
+			openQuorum[q] = startBlock
+		}
+		added := addedToQuorum[op]
+		removed := removedFromQuorum[op]
+		if eventErr := validateQuorumEvents(added, removed, startBlock, endBlock); eventErr != nil {
+			return nil, eventErr
+		}
+		i, j := 0, 0
+		for i < len(added) && j < len(removed) {
+			// TODO(jianoaix): Having quorum addition and removal in the same block is a valid case.
+			// Come up a followup fix to handle this special case.
+			if added[i].BlockNumber == removed[j].BlockNumber {
```

Comment:
Yeah we need to consider the transaction indexes of the transactions or the log index if they happened in the same transaction

Reply:
The idea in my mind was to make a RPC call to the eth node to get the quorums after applying all txns in a block, so we could avoid having to know the details *inside* a block. Quorum opt-in/out are not frequent events, even more so the opt-in/out happening in a same block, so should be small cost to handle.